stan/math/prim/fun/select.hpp

```diff
@@ -0,0 +1,176 @@
+#ifndef STAN_MATH_PRIM_FUN_SELECT_HPP
+#define STAN_MATH_PRIM_FUN_SELECT_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/functor/apply_scalar_binary.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the second argument if the first argument is true
+ * and otherwise return the third argument.
+ *
+ * <code>select(c, y1, y0) = c ? y1 : y0</code>.
+ *
+ * @tparam T_true type of the true argument
+ * @tparam T_false type of the false argument
+ * @param c Boolean condition value.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <typename T_true, typename T_false,
+          typename ReturnT = return_type_t<T_true, T_false>,
+          require_all_stan_scalar_t<T_true, T_false>* = nullptr>
+inline ReturnT select(const bool c, const T_true y_true,
+                      const T_false y_false) {
+  return c ? ReturnT(y_true) : ReturnT(y_false);
+}
+
+/**
+ * Return the second argument if the first argument is true
+ * and otherwise return the third argument. Eigen expressions are
+ * evaluated so that the return type is the same for both branches.
+ *
+ * Both containers must have the same plain type. The scalar type
+ * of the return is determined by the return_type_t<> type trait.
+ *
+ * Overload for use with two containers.
+ *
+ * @tparam T_true type of the true argument
+ * @tparam T_false type of the false argument
+ * @param c Boolean condition value.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <
+    typename T_true, typename T_false,
+    typename T_return = return_type_t<T_true, T_false>,
+    typename T_true_plain = promote_scalar_t<T_return, plain_type_t<T_true>>,
+    typename T_false_plain = promote_scalar_t<T_return, plain_type_t<T_false>>,
+    require_all_container_t<T_true, T_false>* = nullptr,
+    require_all_same_t<T_true_plain, T_false_plain>* = nullptr>
+inline T_true_plain select(const bool c, const T_true y_true,
+                           const T_false y_false) {
+  check_matching_dims("select", "y_true", y_true, "y_false", y_false);
+  return c ? T_true_plain(y_true) : T_true_plain(y_false);
+}
```

Comment:
If these are containers should they be passed as references like 

```suggestion
inline T_true_plain select(const bool c, T_true&& y_true,
                           T_false&& y_false) {
  check_matching_dims("select", "y_true", y_true, "y_false", y_false);
  return c ? T_true_plain(std::forward<T_true>(y_true)) : T_true_plain(std::forward<T_false>(y_false));
}
```